File: planmydinner_addon/api/routine.py
"""
Pasti fissi (colazione e spuntini) con logica opt-out.

Si definiscono una volta per profilo; i totali giornalieri li assumono
consumati ogni giorno (slot con default_on) senza bisogno di registrarli.
Si interviene solo per le eccezioni: "saltato oggi" oppure un pasto diverso
(via flusso mensa con meal_type = slot). Lo slot "dopo_cena" è opt-in:
conta solo quando lo si segna (es. gelato).
"""
import json
import logging
import uuid
from datetime import date
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel as _BaseModel
from sqlalchemy.orm import Session

from ..database import get_db, CandidateRecipe, ConsumedEntry
from ..nutrition import compute_recipe_nutrition

_LOGGER = logging.getLogger(__name__)

router = APIRouter(prefix="/routine", tags=["routine"])

# Slot supportati (ordine = ordine di visualizzazione)
SLOTS = {
    "colazione":          {"label": "Colazione",            "icon": "☕", "default_on": True},
    "spuntino_mattina":   {"label": "Spuntino mattina",     "icon": "🍎", "default_on": True},
    "merenda_pomeriggio": {"label": "Merenda pomeriggio",   "icon": "🥪", "default_on": True},
    "merenda_tardo":      {"label": "Merenda tardo pom.",   "icon": "🍌", "default_on": True},
    "dopo_cena":          {"label": "Dopo cena",            "icon": "🍨", "default_on": False},
}


class RoutineIngredient(_BaseModel):
    name: str
    food_group: str = "altro"
    grams: float


class RoutineMealSave(_BaseModel):
    profile_id: str
    name: str
    ingredients: List[RoutineIngredient]
    default_on: Optional[bool] = None   # None = default dello slot


def _routine_data(cand: CandidateRecipe) -> Optional[dict]:
    data = cand.recipe_data
    if isinstance(data, str):
        try:
            data = json.loads(data)
        except Exception:
            return None
    if not isinstance(data, dict):
        return None
    tags = data.get("tags") or {}
    return data if "true" in (tags.get("routine") or []) else None


def get_routine_meals(db: Session, profile_id: str) -> dict:
    """slot -> {id, name, content, default_on} per il profilo (usato anche da /integration/summary)."""
    out = {}
    for cand in db.query(CandidateRecipe).filter(CandidateRecipe.status == "draft_structured").all():
        data = _routine_data(cand)
        if not data or data.get("profile_id") != profile_id:
            continue
        slot = data.get("meal_slot")
        if slot in SLOTS:
            out[slot] = {
                "id": cand.id,
                "name": data.get("name"),
                "content": data.get("content", []),
                "default_on": bool(data.get("default_on", SLOTS[slot]["default_on"])),
            }
    return out


def _day_entries(db: Session, profile_id: str, iso_date: str, slot: str):
    return db.query(ConsumedEntry).filter(
        ConsumedEntry.profile_id == profile_id,
        ConsumedEntry.date == iso_date,
        ConsumedEntry.meal_type == slot,
    ).all()


def _slot_today_state(entries, meal) -> str:
    """'skipped' | 'logged' (qualcosa registrato) | 'assumed' (default) | 'off'."""
    if any(e.type == "skipped" for e in entries):
        return "skipped"
    if any(e.type != "skipped" for e in entries):
        return "logged"
    return "assumed" if meal and meal["default_on"] else "off"


@router.get("/")
def list_routine(request: Request, profile_id: str, target_date: Optional[date] = None,
                 db: Session = Depends(get_db)):
    """Slot con pasto fisso definito (se c'è), macro e stato del giorno."""
    gw = getattr(request.app.state, "llm_gateway", None)
    meals = get_routine_meals(db, profile_id)
    iso = (target_date or date.today()).isoformat()
    out = []
    for slot, info in SLOTS.items():
        meal = meals.get(slot)
        nutrition = None
        if meal:
            try:
                nutrition = compute_recipe_nutrition(meal["content"], profile_id, llm_gateway=gw)
            except Exception:
                pass
        entries = _day_entries(db, profile_id, iso, slot)
        ingredients = []
        for ing in (meal or {}).get("content", []):
            if isinstance(ing, dict):
                qty = (ing.get("quantities") or {}).get(profile_id) or {}
                ingredients.append({"name": ing.get("name"), "food_group": ing.get("food_group"),
                                    "grams": qty.get("grams_equiv") or qty.get("qty") or 0})
        out.append({
            "slot": slot,
            "label": info["label"],
            "icon": info["icon"],
            "defined": meal is not None,
            "name": (meal or {}).get("name"),
            "ingredients": ingredients,
            "default_on": meal["default_on"] if meal else info["default_on"],
            "nutrition": nutrition,
            "today": _slot_today_state(entries, meal),
        })
    return {"date": iso, "slots": out}


@router.put("/{slot}")
def save_routine_meal(slot: str, body: RoutineMealSave, db: Session = Depends(get_db)):
    """Definisce (o aggiorna) il pasto fisso di uno slot per un profilo."""
    if slot not in SLOTS:
        raise HTTPException(status_code=404, detail=f"Slot sconosciuto: {slot}")
    if not body.ingredients:
        raise HTTPException(status_code=422, detail="Serve almeno un ingrediente.")

    content = [
        {
            "name": ing.name,
            "food_group": ing.food_group,
            "quantities": {body.profile_id: {"qty": float(ing.grams), "unit": "g", "grams_equiv": float(ing.grams)}},
        }
        for ing in body.ingredients if ing.grams > 0
    ]
    default_on = SLOTS[slot]["default_on"] if body.default_on is None else bool(body.default_on)
    recipe_data = {
        "name": body.name.strip(),
        "description": f"Pasto fisso: {SLOTS[slot]['label']}",
        "is_composed_dish": False,
        "content": content,
        "steps": [],
        "total_time_minutes": 0,
        "difficulty": "sconosciuto",
        "tags": {"routine": ["true"]},
        "meal_slot": slot,
        "profile_id": body.profile_id,
        "default_on": default_on,
    }

    existing = get_routine_meals(db, body.profile_id).get(slot)
    if existing:
        cand = db.query(CandidateRecipe).filter(CandidateRecipe.id == existing["id"]).first()
        cand.recipe_data = recipe_data
        from sqlalchemy.orm.attributes import flag_modified
        flag_modified(cand, "recipe_data")
    else:
        cand = CandidateRecipe(id=str(uuid.uuid4()), status="draft_structured", usage_count=0,
                               recipe_data=recipe_data)
        db.add(cand)
    db.commit()
    db.refresh(cand)
    return {"slot": slot, "recipe_id": cand.id, "default_on": default_on}


@router.delete("/{slot}")
def delete_routine_meal(slot: str, profile_id: str, db: Session = Depends(get_db)):
    existing = get_routine_meals(db, profile_id).get(slot)
    if not existing:
        raise HTTPException(status_code=404, detail="Nessun pasto fisso per questo slot.")
    cand = db.query(CandidateRecipe).filter(CandidateRecipe.id == existing["id"]).first()
    db.delete(cand)
    db.commit()
    return {"deleted": slot}


@router.post("/{slot}/skip")
def toggle_skip(slot: str, profile_id: str, meal_date: str, db: Session = Depends(get_db)):
    """Toggle 'saltato oggi': crea (o rimuove) l'eccezione per il giorno."""
    if slot not in SLOTS:
        raise HTTPException(status_code=404, detail=f"Slot sconosciuto: {slot}")
    entries = [e for e in _day_entries(db, profile_id, meal_date, slot) if e.type == "skipped"]
    if entries:
        for e in entries:
            db.delete(e)
        db.commit()
        return {"slot": slot, "date": meal_date, "state": "assumed"}
    db.add(ConsumedEntry(id=str(uuid.uuid4()), profile_id=profile_id, date=meal_date,
                         meal_type=slot, type="skipped"))
    db.commit()
    return {"slot": slot, "date": meal_date, "state": "skipped"}


@router.post("/{slot}/log")
def toggle_log(slot: str, profile_id: str, meal_date: str, db: Session = Depends(get_db)):
    """Toggle registrazione del pasto fisso (per gli slot opt-in, es. gelato dopo cena)."""
    meal = get_routine_meals(db, profile_id).get(slot)
    if not meal:
        raise HTTPException(status_code=404, detail="Nessun pasto fisso definito per questo slot.")
    entries = [e for e in _day_entries(db, profile_id, meal_date, slot)
               if e.type != "skipped" and e.consumed_recipe_id == meal["id"]]
    if entries:
        for e in entries:
            db.delete(e)
        db.commit()
        return {"slot": slot, "date": meal_date, "state": "off"}
    db.add(ConsumedEntry(id=str(uuid.uuid4()), profile_id=profile_id, date=meal_date,
                         meal_type=slot, type="override", consumed_recipe_id=meal["id"],
                         override_details={"free_text_name": meal["name"], "notes": "routine"}))
    db.commit()
    return {"slot": slot, "date": meal_date, "state": "logged"}

File: planmydinner_addon/frontend/routine.js
import { defineComponent } from 'vue';
import MensaModal from './mensa.js?v=5';

/**
 * Strip "Colazione & spuntini" (vista Oggi): pasti fissi con logica opt-out.
 * Gli slot definiti vengono assunti consumati ogni giorno; qui si gestiscono
 * solo le eccezioni (saltato / diverso / opt-in tipo gelato dopo cena) e la
 * definizione dei pasti tipo.
 */
const RoutineStrip = defineComponent({
    name: 'RoutineStrip',
    inject: ['toast'],
    components: { MensaModal },
    props: {
        profileId: { type: String, required: true },
        mealDate: { type: String, required: true },
    },
    emits: ['changed'],
    template: `
        <div class="card routine-strip">
            <div style="display:flex;align-items:center;gap:8px;">
                <strong style="font-size:14px;">☕ Colazione & spuntini</strong>
                <span v-if="definedSlots.length" class="hint">~{{ assumedKcal }} kcal oggi</span>
                <button @click="editorOpen = true" class="btn-sm btn-secondary" style="margin-left:auto;">⚙️</button>
            </div>

            <div v-if="loading" class="hint" style="margin-top:6px;">Caricamento…</div>

            <div v-else-if="!definedSlots.length" class="hint" style="margin-top:6px;">
                Definisci una volta la tua colazione e le merende tipo: verranno contate
                nei totali ogni giorno senza doverle registrare.
                <button @click="editorOpen = true" class="btn-sm btn-primary" style="margin-left:6px;">Definisci</button>
            </div>

            <div v-else style="display:flex;flex-direction:column;gap:6px;margin-top:8px;">
                <div v-for="s in definedSlots" :key="s.slot" class="routine-row">
                    <span style="font-size:15px;">{{ s.icon }}</span>
                    <span class="routine-row__name">
                        {{ s.name }}
                        <span v-if="s.nutrition" class="hint">~{{ Math.round(s.nutrition.kcal) }} kcal</span>
                    </span>
                    <!-- Stato / azioni del giorno -->
                    <template v-if="s.default_on">
                        <button v-if="s.today === 'assumed'" @click="toggleSkip(s)" class="chip chip--on" title="Tocca se oggi l'hai saltata">✓</button>
                        <button v-else-if="s.today === 'skipped'" @click="toggleSkip(s)" class="chip chip--off" title="Segnata come saltata — tocca per ripristinare">✗ saltata</button>
                        <span v-else class="chip chip--logged" title="Registrato un pasto diverso">✎ diversa</span>
                    </template>
                    <template v-else>
                        <button v-if="s.today === 'off'" @click="toggleLog(s)" class="chip chip--add" title="Segna per oggi">＋</button>
                        <button v-else-if="s.today === 'logged'" @click="toggleLog(s)" class="chip chip--on" title="Segnata — tocca per togliere">✓</button>
                        <button v-else-if="s.today === 'skipped'" @click="toggleSkip(s)" class="chip chip--off">✗</button>
                    </template>
                    <button @click="mensaSlot = s" class="btn-sm btn-secondary" title="Oggi qualcosa di diverso (foto/descrizione)">≠</button>
                </div>
            </div>

            <!-- Pasto diverso per uno slot -->
            <mensa-modal v-if="mensaSlot"
                         :profile-id="profileId"
                         :meal-type="mensaSlot.slot"
                         :meal-label="mensaSlot.label"
                         :meal-date="mealDate"
                         @close="mensaSlot = null"
                         @saved="onChanged" />

            <!-- Editor pasti fissi -->
            <div v-if="editorOpen" class="modal-overlay" @click.self="closeEditor">
                <div class="modal">
                    <h3>☕ Pasti fissi</h3>
                    <p class="hint" style="margin:0 0 10px;">
                        Vengono contati automaticamente nei totali di ogni giorno.
                        "Dopo cena" è l'eccezione: conta solo quando lo segni.
                    </p>

                    <div v-for="s in slots" :key="s.slot" style="border:1px solid var(--border);border-radius:10px;padding:10px 12px;margin-bottom:8px;">
                        <div style="display:flex;align-items:center;gap:8px;cursor:pointer;" @click="toggleEdit(s)">
                            <span>{{ s.icon }}</span>
                            <strong style="flex:1;font-size:13.5px;">{{ s.label }}</strong>
                            <span v-if="s.defined" class="hint">{{ s.name }} · ~{{ s.nutrition ? Math.round(s.nutrition.kcal) : '?' }} kcal</span>
                            <span v-else class="hint">non definito</span>
                            <span style="color:var(--text-3);">{{ editSlot === s.slot ? '▲' : '✏️' }}</span>
                        </div>

                        <div v-if="editSlot === s.slot" style="margin-top:10px;display:flex;flex-direction:column;gap:8px;">
                            <div style="display:flex;gap:8px;">
                                <input v-model="editForm.description"
                                       placeholder="es. caffè, 2 fette biscottate con marmellata, yogurt greco"
                                       :disabled="analyzing" @keyup.enter="analyzeDescription">
                                <button @click="analyzeDescription" class="btn-primary btn-sm"
                                        :disabled="analyzing || editForm.description.trim().length < 3" style="flex-shrink:0;">
                                    {{ analyzing ? '…' : '✨' }}
                                </button>
                            </div>
                            <input v-model="editForm.name" placeholder="Nome (es. Colazione tipo)">
                            <div v-for="(ing, idx) in editForm.ingredients" :key="idx"
                                 style="display:grid;grid-template-columns:1fr 80px auto;gap:8px;align-items:center;">
                                <input v-model="ing.name">
                                <input v-model.number="ing.grams" type="number" min="0" step="5">
                                <button @click="editForm.ingredients.splice(idx, 1)" class="btn-secondary btn-sm">✕</button>
                            </div>
                            <button @click="editForm.ingredients.push({name: '', food_group: 'altro', grams: 50})"
                                    class="btn-secondary btn-sm" style="align-self:flex-start;">+ ingrediente</button>
                            <label class="toggle-label">
                                <input type="checkbox" v-model="editForm.default_on" style="width:auto;">
                                Conta ogni giorno automaticamente
                            </label>
                            <div style="display:flex;gap:8px;">
                                <button @click="saveSlot(s)" class="btn-primary btn-sm"
                                        :disabled="saving || !editForm.name || !editForm.ingredients.length">
                                    {{ saving ? 'Salvataggio…' : '💾 Salva' }}
                                </button>
                                <button v-if="s.defined" @click="deleteSlot(s)" class="btn-danger btn-sm" style="margin-left:auto;">🗑️</button>
                            </div>
                            <div v-if="editError" class="error">{{ editError }}</div>
                        </div>
                    </div>

                    <button @click="closeEditor" class="btn-secondary" style="margin-top:8px;">Chiudi</button>
                </div>
            </div>
        </div>
    `,
    data() {
        return {
            slots: [],
            loading: false,
            editorOpen: false,
            editSlot: null,
            editForm: { description: '', name: '', ingredients: [], default_on: true },
            analyzing: false,
            saving: false,
            editError: null,
            mensaSlot: null,
        };
    },
    computed: {
        definedSlots() { return this.slots.filter(s => s.defined); },
        assumedKcal() {
            return Math.round(this.definedSlots.reduce((sum, s) => {
                if (!s.nutrition) return sum;
                const counts = (s.default_on && s.today !== 'skipped' && s.today !== 'logged')
                    || (!s.default_on && s.today === 'logged');
                return counts ? sum + s.nutrition.kcal : sum;
            }, 0));
        },
    },
    mounted() { this.load(); },
    methods: {
        async load() {
            this.loading = true;
            try {
                const params = new URLSearchParams({ profile_id: this.profileId, target_date: this.mealDate });
                const resp = await window.apiFetch('/routine/?' + params);
                this.slots = resp.ok ? (await resp.json()).slots : [];
            } catch (_) {
                this.slots = [];
            } finally {
                this.loading = false;
            }
        },
        async onChanged() {
            this.mensaSlot = null;
            await this.load();
            this.$emit('changed');
        },
        async toggleSkip(s) {
            try {
                const params = new URLSearchParams({ profile_id: this.profileId, meal_date: this.mealDate });
                const resp = await window.apiFetch(`/routine/${s.slot}/skip?` + params, { method: 'POST' });
                if (!resp.ok) throw new Error(await resp.text());
                const r = await resp.json();
                this.toast.add(r.state === 'skipped' ? `${s.label} saltata oggi` : `${s.label} ripristinata`, 'info');
                await this.onChanged();
            } catch (e) {
                this.toast.add('Errore: ' + e.message, 'error');
            }
        },
        async toggleLog(s) {
            try {
                const params = new URLSearchParams({ profile_id: this.profileId, meal_date: this.mealDate });
                const resp = await window.apiFetch(`/routine/${s.slot}/log?` + params, { method: 'POST' });
                if (!resp.ok) throw new Error(await resp.text());
                const r = await resp.json();
                this.toast.add(r.state === 'logged' ? `${s.icon} ${s.name} segnato!` : `${s.label} tolta`, 'success');
                await this.onChanged();
            } catch (e) {
                this.toast.add('Errore: ' + e.message, 'error');
            }
        },
        toggleEdit(s) {
            if (this.editSlot === s.slot) { this.editSlot = null; return; }
            this.editSlot = s.slot;
            this.editError = null;
            this.editForm = {
                description: '',
                name: s.name || '',
                ingredients: (s.ingredients || []).map(i => ({ ...i })),
                default_on: s.default_on,
            };
        },
        async analyzeDescription() {
            const description = this.editForm.description.trim();
            if (description.length < 3) return;
            this.analyzing = true;
            this.editError = null;
            try {
                const params = new URLSearchParams({ profile_id: this.profileId });
                const resp = await window.apiFetch('/consumed-entries/text/analyze?' + params, {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify({ description }),
                });
                if (!resp.ok) {
                    const err = await resp.json().catch(() => ({}));
                    throw new Error(err.detail || 'Analisi fallita.');
                }
                const proposal = await resp.json();
                this.editForm.name = this.editForm.name || proposal.name;
                this.editForm.ingredients = proposal.ingredients;
            } catch (e) {
                this.editError = e.message;
            } finally {
                this.analyzing = false;
            }
        },
        async saveSlot(s) {
            this.saving = true;
            this.editError = null;
            try {
                const body = {
                    profile_id: this.profileId,
                    name: this.editForm.name,
                    ingredients: this.editForm.ingredients.filter(i => i.name && i.grams > 0),
                    default_on: this.editForm.default_on,
                };
                const resp = await window.apiFetch(`/routine/${s.slot}`, {
                    method: 'PUT',
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify(body),
                });
                if (!resp.ok) throw new Error(await resp.text());
                this.toast.add(`${s.label} salvata!`, 'success');
                this.editSlot = null;
                await this.load();
                this.$emit('changed');
            } catch (e) {
                this.editError = 'Errore: ' + e.message;
            } finally {
                this.saving = false;
            }
        },
        async deleteSlot(s) {
            if (!confirm(`Rimuovere "${s.name}" da ${s.label}?`)) return;
            try {
                const params = new URLSearchParams({ profile_id: this.profileId });
                const resp = await window.apiFetch(`/routine/${s.slot}?` + params, { method: 'DELETE' });
                if (!resp.ok) throw new Error(await resp.text());
                this.editSlot = null;
                await this.load();
                this.$emit('changed');
            } catch (e) {
                this.toast.add('Errore: ' + e.message, 'error');
            }
        },
        closeEditor() {
            this.editorOpen = false;
            this.editSlot = null;
        },
    },
});

export default RoutineStrip;
